api/v1alpha1/nicclusterpolicy_webhook.go

```diff
@@ -0,0 +1,297 @@
+limitations under the License.
+*/
+
+package v1alpha1
+
+import (
+	"encoding/json"
+	"math/big"
+	"regexp"
+	"strings"
+
+	"github.com/xeipuuv/gojsonschema"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/util/validation/field"
+	ctrl "sigs.k8s.io/controller-runtime"
+	logf "sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/webhook"
+)
+
+// log is for logging in this package.
+var nicclusterpolicylog = logf.Log.WithName("nicclusterpolicy-resource")
+
+func (r *NicClusterPolicy) SetupWebhookWithManager(mgr ctrl.Manager) error {
+	nicclusterpolicylog.Info("Nic cluster policy webhook admission controller")
+	return ctrl.NewWebhookManagedBy(mgr).
+		For(r).
+		Complete()
+}
+
+// TODO(user): change verbs to "verbs=create;update;delete" if you want to enable deletion validation.
+//nolint:lll
+//+kubebuilder:webhook:path=/validate-mellanox-com-v1alpha1-nicclusterpolicy,mutating=false,failurePolicy=fail,sideEffects=None,groups=mellanox.com,resources=nicclusterpolicies,verbs=create;update,versions=v1alpha1,name=vnicclusterpolicy.kb.io,admissionReviewVersions=v1
+
+var _ webhook.Validator = &NicClusterPolicy{}
+
+// ValidateCreate implements webhook.Validator so a webhook will be registered for the type
+func (r *NicClusterPolicy) ValidateCreate() error {
+	nicclusterpolicylog.Info("validate create", "name", r.Name)
+	return r.validateNicClusterPolicy()
+}
+
+// ValidateUpdate implements webhook.Validator so a webhook will be registered for the type
+func (r *NicClusterPolicy) ValidateUpdate(_ runtime.Object) error {
+	nicclusterpolicylog.Info("validate update", "name", r.Name)
+	return r.validateNicClusterPolicy()
+}
+
+// ValidateDelete implements webhook.Validator so a webhook will be registered for the type
+func (r *NicClusterPolicy) ValidateDelete() error {
+	nicclusterpolicylog.Info("validate delete", "name", r.Name)
+
+	// TODO(user): fill in your validation logic upon object deletion.
+	return nil
+}
+
+/*
+We validate here
+*/
+func (r *NicClusterPolicy) validateNicClusterPolicy() error {
+	var allErrs field.ErrorList
+	// Validate IBKubernetes
+	ibKubernetes := r.Spec.IBKubernetes
+	if ibKubernetes != nil {
+		allErrs = append(allErrs, ibKubernetes.validate(field.NewPath("spec").Child("ibKubernetes"))...)
+	}
+
+	// Validate OFEDDriverSpec
+	ofedDriver := r.Spec.OFEDDriver
+	if ofedDriver != nil {
+		allErrs = append(allErrs, ofedDriver.validateVersion(field.NewPath("spec").Child("ofedDriver"))...)
+	}
+	// Validate RdmaSharedDevicePlugin
+	rdmaSharedDevicePlugin := r.Spec.RdmaSharedDevicePlugin
+	if rdmaSharedDevicePlugin != nil {
+		allErrs = append(allErrs, r.Spec.RdmaSharedDevicePlugin.validateRdmaSharedDevicePlugin(
+			field.NewPath("spec").Child("rdmaSharedDevicePlugin"))...)
+	}
+	// Validate SriovDevicePlugin
+	SriovNetworkDevicePlugin := r.Spec.SriovDevicePlugin
```

Comment:
why the variable name starts with uppercase S?

Reply:
Done